src/model_server/model_server_manager.cpp

```diff
@@ -0,0 +1,135 @@
+#include "model_server/model_server_manager.h"
+
+#include <csignal.h>
+#include <sys/wait.h>
+
+#include <filesystem>
+#include <thread>  // NOLINT
+
+#include "common/json.h"
+#include "loggers/model_logger.h"
+#include "messenger/connection_destination.h"
+#include "messenger/messenger.h"
+
+namespace noisepage::model {
+
+/**
+ * This initializes a connection to the model by openning up a zmq connection
+ * @param messenger
+ * @return A ConnectionId that should be used only to the calling thread
+ */
+common::ManagedPointer<messenger::ConnectionRouter> ListenAndMakeConnection(
+    const common::ManagedPointer<messenger::Messenger> &messenger, const std::string &ipc_path,
+    messenger::CallbackFn model_server_logic) {
+  // Create an IPC connection that the Python process will talk to.
+  auto destination = messenger::ConnectionDestination::MakeIPC(MODEL_TARGET_NAME, ipc_path);
+
+  // Listen for the connection
+  messenger->ListenForConnection(destination, MODEL_CONN_ID_NAME, std::move(model_server_logic));
+  while (true) {
+    try {
+      return messenger->GetConnectionRouter(MODEL_CONN_ID_NAME);
+    } catch (std::exception &e) {
+      ::sleep(1);
+    }
+  }
+}
+
+}  // namespace noisepage::model
+
+namespace noisepage::model {
+
+void ModelServerManager::ModelServerHandler(common::ManagedPointer<messenger::Messenger> messenger,
+                                            std::string_view sender_id, std::string_view message, uint64_t recv_cb_id) {
+  // TODO(ricky): Currently just echo the results. No continuation
+  MESSENGER_LOG_TRACE("[PID={},SEND_ID={},RECV_ID={}] Messenger RECV: {}", ::getpid(), message, sender_id, recv_cb_id);
+  (void)messenger;
+}
+
+ModelServerManager::ModelServerManager(const std::string &model_bin,
+                                       const common::ManagedPointer<messenger::Messenger> &messenger)
+    : messenger_(messenger),
+      router_(ListenAndMakeConnection(messenger, MODEL_IPC_PATH, ModelServerHandler)),
+      thd_(std::thread([this, model_bin] {
+        while (!shut_down_) {
+          this->StartModelServer(model_bin);
+        }
+      })) {}
+
+void ModelServerManager::StartModelServer(const std::string &model_path) {
```

Comment:
nit: `model_path` to `model_server_path`. Change the naming to be explicit in other places as well.